packages/ui/src/components/va-counter/VaCounter.vue

```diff
@@ -0,0 +1,368 @@
+import { useEmitProxy } from '../../composables/useEmitProxy'
+import { useFocus, useFocusEmits } from '../../composables/useFocus'
+import { useStatefulProps, useStateful } from '../../composables/useStateful'
+import { useColor } from '../../composables/useColor'
+import VaInputWrapper from '../va-input/components/VaInputWrapper.vue'
+import VaIcon from '../va-icon/VaIcon.vue'
+import VaButton from '../va-button/VaButton.vue'
+import { omit, pick } from 'lodash-es'
+
+const { createEmits: createInputEmits, createListeners: createInputListeners } = useEmitProxy(
+  ['change'],
+)
+
+const { createEmits: createFieldEmits, createListeners: createFieldListeners } = useEmitProxy([
+  { listen: 'click-prepend', emit: 'click:decrease-button' },
+  { listen: 'click-append', emit: 'click:increase-button' },
+  { listen: 'click-prepend-inner', emit: 'click:decrease-icon' },
+  { listen: 'click-append-inner', emit: 'click:increase-icon' },
+])
+
+export default defineComponent({
+  name: 'VaCounter',
+
+  components: { VaInputWrapper, VaIcon, VaButton },
+
+  props: {
+    ...useFormProps,
+    ...useStatefulProps,
+    // input
+    modelValue: { type: [String, Number], default: '' },
+    manualInput: { type: Boolean, default: false },
+    stateful: { type: Boolean, default: false },
+    min: { type: Number, default: 1 },
+    max: { type: Number, default: undefined },
+    step: { type: Number, default: 1 },
+    label: { type: String, default: '' },
+    // hint
+    messages: { type: [Array, String] as PropType<string[] | string>, default: () => [] },
+    // style
+    width: { type: String, default: '154px' },
+    color: { type: String, default: 'primary' },
+    outline: { type: Boolean, default: undefined },
+    bordered: { type: Boolean, default: undefined },
+    // icons & buttons
+    increaseIcon: { type: String, default: 'add' },
+    decreaseIcon: { type: String, default: 'remove' },
+    buttons: { type: Boolean, default: false },
+    margins: { type: [String, Number], default: '4px' },
+    flat: { type: Boolean as PropType<boolean>, default: true },
+    rounded: { type: Boolean as PropType<boolean>, default: false },
+    textColor: { type: String as PropType<string | undefined>, default: undefined },
+  },
+
+  emits: [
+    'update:modelValue',
+    ...createInputEmits(),
+    ...createFieldEmits(),
+    ...useFocusEmits,
+  ],
+
+  inheritAttrs: false,
+
+  setup (props, { emit, attrs }) {
+    const input = ref<HTMLInputElement | HTMLDivElement | undefined>()
+
+    const { isFocused, focus, blur } = useFocus(input, emit)
+
+    const { valueComputed } = useStateful(props, emit)
+
+    const onManualInput = (counterValue: string) => {
+      valueComputed.value = counterValue
+    }
+
+    const getRoundDownWithStep = (value: number | string) => {
+      return props.min + props.step * Math.floor((+value - props.min) / props.step)
+    }
+
+    const onManualChange = (counterValue: number | string) => {
```

Comment:
Not a fun of 'onEvent' names. There's no way to guess what this handler does.

Reply:
done